lib/common/bits.h

```diff
@@ -0,0 +1,244 @@
+            __assume(0);
+        }
+#   elif defined(__GNUC__) && (__GNUC__ >= 3)
+        return (unsigned)__builtin_ctz(val);
+#   elif defined(__ICCARM__)    /* IAR Intrinsic */
+        return __CTZ(val);
+#   else
+        return ZSTD_countTrailingZeros32_fallback(val);
+#   endif
+}
+
+MEM_STATIC unsigned ZSTD_countTrailingZeros64_fallback(U64 val)
+{
+    static const int DeBruijnBytePos[64] = {  0,  1,  2,  7,  3, 13,  8, 19,
+                                              4, 25, 14, 28,  9, 34, 20, 56,
+                                              5, 17, 26, 54, 15, 41, 29, 43,
+                                             10, 31, 38, 35, 21, 45, 49, 57,
+                                             63,  6, 12, 18, 24, 27, 33, 55,
+                                             16, 53, 40, 42, 30, 37, 44, 48,
+                                             62, 11, 23, 32, 52, 39, 36, 47,
+                                             61, 22, 51, 46, 60, 50, 59, 58 };
+    return DeBruijnBytePos[((U64)((val & -(long long)val) * 0x0218A392CDABBD3FULL)) >> 58];
+}
+
+MEM_STATIC unsigned ZSTD_countTrailingZeros64(U64 val)
+{
+    assert(val != 0);
+#   if defined(_MSC_VER) && defined(_WIN64)
+#       if STATIC_BMI2
+            return _tzcnt_u64(val);
+#       else
+            if (val != 0) {
+                unsigned long r;
+                _BitScanForward64(&r, val);
+                return (unsigned)r;
+            } else {
+                /* Should not reach this code path */
+                __assume(0);
+            }
+#       endif
+#   elif defined(__GNUC__) && (__GNUC__ >= 4)
+        if (MEM_32bits()) {
+            U32 mostSignificantWord = (U32)(val >> 32);
+            U32 leastSignificantWord = (U32)val;
+            if (leastSignificantWord == 0) {
+                return 32 + (unsigned)__builtin_ctz(mostSignificantWord);
+            } else {
+                return (unsigned)__builtin_ctz(leastSignificantWord);
+            }
+        } else {
+            return (unsigned)__builtin_ctzll(val);
+        }
+#   else
+        return ZSTD_countTrailingZeros64_fallback(val);
+#   endif
+}
+
+MEM_STATIC unsigned ZSTD_NbCommonBytes_fallback(size_t val)
+{
+    if (MEM_isLittleEndian()) {
+        if (MEM_64bits()) {
+            static const int DeBruijnBytePos[64] = { 0, 0, 0, 0, 0, 1, 1, 2,
+                                                     0, 3, 1, 3, 1, 4, 2, 7,
+                                                     0, 2, 3, 6, 1, 5, 3, 5,
+                                                     1, 3, 4, 4, 2, 5, 6, 7,
+                                                     7, 0, 1, 2, 3, 3, 4, 6,
+                                                     2, 6, 5, 5, 3, 4, 5, 6,
+                                                     7, 1, 2, 4, 6, 4, 4, 5,
+                                                     7, 2, 6, 5, 7, 6, 7, 7 };
+            return DeBruijnBytePos[((U64)((val & -(long long)val) * 0x0218A392CDABBD3FULL)) >> 58];
```

Comment:
```
const U64 m = 0x0101010101010101ULL;
val ^= val - 1;
return (unsigned)(((U64)((val & (m - 1)) * m)) >> 56);
```

Reply:
@aqrit Can we address this in a separate PR?

Reply:
> Then you can implement ZSTD_NbCommonBytes() without any #ifdef code or fallbacks.
The only difference is that the software fallback would need one extra `>> 3` instruction. But I don't think that is a big deal.

terrelln is advocating refactoring away this function? So it seem like a good time to being it up :-(

Reply:
@aqrit is there a scenario you care about that uses this fallback code?

Reply:
no.